// square/dist/esm/models/terminalAction.js
import { lazy, nullable, object, optional, string } from '../schema';
import { deviceMetadataSchema } from './deviceMetadata';
import { receiptOptionsSchema } from './receiptOptions';
import { saveCardOptionsSchema } from './saveCardOptions';
export const terminalActionSchema = object({
    id: ['id', optional(string())],
    deviceId: ['device_id', optional(nullable(string()))],
    deadlineDuration: ['deadline_duration', optional(nullable(string()))],
    status: ['status', optional(string())],
    cancelReason: ['cancel_reason', optional(string())],
    createdAt: ['created_at', optional(string())],
    updatedAt: ['updated_at', optional(string())],
    appId: ['app_id', optional(string())],
    type: ['type', optional(string())],
    saveCardOptions: [
        'save_card_options',
        optional(lazy(() => saveCardOptionsSchema)),
    ],
    receiptOptions: [
        'receipt_options',
        optional(lazy(() => receiptOptionsSchema)),
    ],
    deviceMetadata: [
        'device_metadata',
        optional(lazy(() => deviceMetadataSchema)),
    ],
});
