// square/dist/cjs/models/listGiftCardActivitiesRequest.js
"use strict";
exports.__esModule = true;
exports.listGiftCardActivitiesRequestSchema = void 0;
var schema_1 = require("../schema");
exports.listGiftCardActivitiesRequestSchema = (0, schema_1.object)({
    giftCardId: ['gift_card_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    type: ['type', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    locationId: ['location_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    beginTime: ['begin_time', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    endTime: ['end_time', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    limit: ['limit', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    cursor: ['cursor', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    sortOrder: ['sort_order', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))]
});
