// square/dist/cjs/models/terminalAction.js
"use strict";
exports.__esModule = true;
exports.terminalActionSchema = void 0;
var schema_1 = require("../schema");
var deviceMetadata_1 = require("./deviceMetadata");
var receiptOptions_1 = require("./receiptOptions");
var saveCardOptions_1 = require("./saveCardOptions");
exports.terminalActionSchema = (0, schema_1.object)({
    id: ['id', (0, schema_1.optional)((0, schema_1.string)())],
    deviceId: ['device_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    deadlineDuration: ['deadline_duration', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    status: ['status', (0, schema_1.optional)((0, schema_1.string)())],
    cancelReason: ['cancel_reason', (0, schema_1.optional)((0, schema_1.string)())],
    createdAt: ['created_at', (0, schema_1.optional)((0, schema_1.string)())],
    updatedAt: ['updated_at', (0, schema_1.optional)((0, schema_1.string)())],
    appId: ['app_id', (0, schema_1.optional)((0, schema_1.string)())],
    type: ['type', (0, schema_1.optional)((0, schema_1.string)())],
    saveCardOptions: [
        'save_card_options',
        (0, schema_1.optional)((0, schema_1.lazy)(function () { return saveCardOptions_1.saveCardOptionsSchema; })),
    ],
    receiptOptions: [
        'receipt_options',
        (0, schema_1.optional)((0, schema_1.lazy)(function () { return receiptOptions_1.receiptOptionsSchema; })),
    ],
    deviceMetadata: [
        'device_metadata',
        (0, schema_1.optional)((0, schema_1.lazy)(function () { return deviceMetadata_1.deviceMetadataSchema; })),
    ]
});
