// square/dist/cjs/models/v1Payment.js
"use strict";
exports.__esModule = true;
exports.v1PaymentSchema = void 0;
var schema_1 = require("../schema");
var device_1 = require("./device");
var v1Money_1 = require("./v1Money");
var v1PaymentItemization_1 = require("./v1PaymentItemization");
var v1PaymentSurcharge_1 = require("./v1PaymentSurcharge");
var v1PaymentTax_1 = require("./v1PaymentTax");
var v1Refund_1 = require("./v1Refund");
var v1Tender_1 = require("./v1Tender");
exports.v1PaymentSchema = (0, schema_1.object)({
    id: ['id', (0, schema_1.optional)((0, schema_1.string)())],
    merchantId: ['merchant_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    createdAt: ['created_at', (0, schema_1.optional)((0, schema_1.string)())],
    creatorId: ['creator_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    device: ['device', (0, schema_1.optional)((0, schema_1.lazy)(function () { return device_1.deviceSchema; }))],
    paymentUrl: ['payment_url', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    receiptUrl: ['receipt_url', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    inclusiveTaxMoney: [
        'inclusive_tax_money',
        (0, schema_1.optional)((0, schema_1.lazy)(function () { return v1Money_1.v1MoneySchema; })),
    ],
    additiveTaxMoney: ['additive_tax_money', (0, schema_1.optional)((0, schema_1.lazy)(function () { return v1Money_1.v1MoneySchema; }))],
    taxMoney: ['tax_money', (0, schema_1.optional)((0, schema_1.lazy)(function () { return v1Money_1.v1MoneySchema; }))],
    tipMoney: ['tip_money', (0, schema_1.optional)((0, schema_1.lazy)(function () { return v1Money_1.v1MoneySchema; }))],
    discountMoney: ['discount_money', (0, schema_1.optional)((0, schema_1.lazy)(function () { return v1Money_1.v1MoneySchema; }))],
    totalCollectedMoney: [
        'total_collected_money',
        (0, schema_1.optional)((0, schema_1.lazy)(function () { return v1Money_1.v1MoneySchema; })),
    ],
    processingFeeMoney: [
        'processing_fee_money',
        (0, schema_1.optional)((0, schema_1.lazy)(function () { return v1Money_1.v1MoneySchema; })),
    ],
    netTotalMoney: ['net_total_money', (0, schema_1.optional)((0, schema_1.lazy)(function () { return v1Money_1.v1MoneySchema; }))],
    refundedMoney: ['refunded_money', (0, schema_1.optional)((0, schema_1.lazy)(function () { return v1Money_1.v1MoneySchema; }))],
    swedishRoundingMoney: [
        'swedish_rounding_money',
        (0, schema_1.optional)((0, schema_1.lazy)(function () { return v1Money_1.v1MoneySchema; })),
    ],
    grossSalesMoney: ['gross_sales_money', (0, schema_1.optional)((0, schema_1.lazy)(function () { return v1Money_1.v1MoneySchema; }))],
    netSalesMoney: ['net_sales_money', (0, schema_1.optional)((0, schema_1.lazy)(function () { return v1Money_1.v1MoneySchema; }))],
    inclusiveTax: [
        'inclusive_tax',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.array)((0, schema_1.lazy)(function () { return v1PaymentTax_1.v1PaymentTaxSchema; })))),
    ],
    additiveTax: [
        'additive_tax',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.array)((0, schema_1.lazy)(function () { return v1PaymentTax_1.v1PaymentTaxSchema; })))),
    ],
    tender: ['tender', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.array)((0, schema_1.lazy)(function () { return v1Tender_1.v1TenderSchema; }))))],
    refunds: ['refunds', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.array)((0, schema_1.lazy)(function () { return v1Refund_1.v1RefundSchema; }))))],
    itemizations: [
        'itemizations',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.array)((0, schema_1.lazy)(function () { return v1PaymentItemization_1.v1PaymentItemizationSchema; })))),
    ],
    surchargeMoney: ['surcharge_money', (0, schema_1.optional)((0, schema_1.lazy)(function () { return v1Money_1.v1MoneySchema; }))],
    surcharges: [
        'surcharges',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.array)((0, schema_1.lazy)(function () { return v1PaymentSurcharge_1.v1PaymentSurchargeSchema; })))),
    ],
    isPartial: ['is_partial', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.boolean)()))]
});
