// square/dist/cjs/models/v1PaymentSurcharge.js
"use strict";
exports.__esModule = true;
exports.v1PaymentSurchargeSchema = void 0;
var schema_1 = require("../schema");
var v1Money_1 = require("./v1Money");
var v1PaymentTax_1 = require("./v1PaymentTax");
exports.v1PaymentSurchargeSchema = (0, schema_1.object)({
    name: ['name', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    appliedMoney: ['applied_money', (0, schema_1.optional)((0, schema_1.lazy)(function () { return v1Money_1.v1MoneySchema; }))],
    rate: ['rate', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    amountMoney: ['amount_money', (0, schema_1.optional)((0, schema_1.lazy)(function () { return v1Money_1.v1MoneySchema; }))],
    type: ['type', (0, schema_1.optional)((0, schema_1.string)())],
    taxable: ['taxable', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.boolean)()))],
    taxes: ['taxes', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.array)((0, schema_1.lazy)(function () { return v1PaymentTax_1.v1PaymentTaxSchema; }))))],
    surchargeId: ['surcharge_id', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))]
});
