// square/dist/cjs/models/v1PaymentItemization.js
"use strict";
exports.__esModule = true;
exports.v1PaymentItemizationSchema = void 0;
var schema_1 = require("../schema");
var v1Money_1 = require("./v1Money");
var v1PaymentDiscount_1 = require("./v1PaymentDiscount");
var v1PaymentItemDetail_1 = require("./v1PaymentItemDetail");
var v1PaymentModifier_1 = require("./v1PaymentModifier");
var v1PaymentTax_1 = require("./v1PaymentTax");
exports.v1PaymentItemizationSchema = (0, schema_1.object)({
    name: ['name', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    quantity: ['quantity', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.number)()))],
    itemizationType: ['itemization_type', (0, schema_1.optional)((0, schema_1.string)())],
    itemDetail: ['item_detail', (0, schema_1.optional)((0, schema_1.lazy)(function () { return v1PaymentItemDetail_1.v1PaymentItemDetailSchema; }))],
    notes: ['notes', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    itemVariationName: ['item_variation_name', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.string)()))],
    totalMoney: ['total_money', (0, schema_1.optional)((0, schema_1.lazy)(function () { return v1Money_1.v1MoneySchema; }))],
    singleQuantityMoney: [
        'single_quantity_money',
        (0, schema_1.optional)((0, schema_1.lazy)(function () { return v1Money_1.v1MoneySchema; })),
    ],
    grossSalesMoney: ['gross_sales_money', (0, schema_1.optional)((0, schema_1.lazy)(function () { return v1Money_1.v1MoneySchema; }))],
    discountMoney: ['discount_money', (0, schema_1.optional)((0, schema_1.lazy)(function () { return v1Money_1.v1MoneySchema; }))],
    netSalesMoney: ['net_sales_money', (0, schema_1.optional)((0, schema_1.lazy)(function () { return v1Money_1.v1MoneySchema; }))],
    taxes: ['taxes', (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.array)((0, schema_1.lazy)(function () { return v1PaymentTax_1.v1PaymentTaxSchema; }))))],
    discounts: [
        'discounts',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.array)((0, schema_1.lazy)(function () { return v1PaymentDiscount_1.v1PaymentDiscountSchema; })))),
    ],
    modifiers: [
        'modifiers',
        (0, schema_1.optional)((0, schema_1.nullable)((0, schema_1.array)((0, schema_1.lazy)(function () { return v1PaymentModifier_1.v1PaymentModifierSchema; })))),
    ]
});
