// @apimatic/schema/es/index.js
export { validateAndMap, validateAndMapXml, validateAndUnmap, validateAndUnmapXml } from './schema.js';
export { array } from './types/array.js';
export { bigint } from './types/bigint.js';
export { boolean } from './types/boolean.js';
export { defaults } from './types/defaults.js';
export { dict } from './types/dict.js';
export { dictWithXmlEntries } from './types/dictWithXmlEntries.js';
export { discriminatedObject } from './types/discriminatedObject.js';
export { lazy } from './types/lazy.js';
export { literal } from './types/literal.js';
export { nullable } from './types/nullable.js';
export { number } from './types/number.js';
export { numberEnum } from './types/numberEnum.js';
export { expandoObject, extendExpandoObject, extendObject, extendStrictObject, object, strictObject } from './types/object.js';
export { optional } from './types/optional.js';
export { string } from './types/string.js';
export { stringEnum } from './types/stringEnum.js';
export { unknown } from './types/unknown.js';
